executor/mem_reader.go

```diff
@@ -0,0 +1,454 @@
+}
+
+func (m *memIndexReader) decodeIndexKeyValue(key, value []byte, tps []*types.FieldType) ([]types.Datum, error) {
+	// this is from indexScanExec decodeIndexKV method.
+	values, b, err := tablecodec.CutIndexKeyNew(key, len(m.index.Columns))
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(b) > 0 {
+		values = append(values, b)
+	} else if len(value) >= 8 {
+		handle, err := decodeHandle(value)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		var handleDatum types.Datum
+		if mysql.HasUnsignedFlag(tps[len(tps)-1].Flag) {
+			handleDatum = types.NewUintDatum(uint64(handle))
+		} else {
+			handleDatum = types.NewIntDatum(handle)
+		}
+		m.handleBytes, err = codec.EncodeValue(m.ctx.GetSessionVars().StmtCtx, m.handleBytes[:0], handleDatum)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		values = append(values, m.handleBytes)
+	}
+
+	ds := make([]types.Datum, 0, len(m.outputOffset))
+	for _, offset := range m.outputOffset {
+		d, err := tablecodec.DecodeColumnValue(values[offset], tps[offset], m.ctx.GetSessionVars().TimeZone)
+		if err != nil {
+			return nil, err
+		}
+		ds = append(ds, d)
+	}
+	return ds, nil
+}
+
+func (m *memIndexReader) decodeIndexHandle(key, value []byte, pkTp *types.FieldType) (int64, error) {
+	_, b, err := tablecodec.CutIndexKeyNew(key, len(m.index.Columns))
+	if err != nil {
+		return 0, errors.Trace(err)
+	}
+	if len(b) > 0 {
+		d, err := tablecodec.DecodeColumnValue(b, pkTp, m.ctx.GetSessionVars().TimeZone)
+		if err != nil {
+			return 0, errors.Trace(err)
+		}
+		return d.GetInt64(), nil
+
+	} else if len(value) >= 8 {
+		return decodeHandle(value)
+	}
+	// Should never execute to here.
+	return 0, errors.Errorf("no handle in index key: %v, value: %v", key, value)
+}
+
+func decodeHandle(data []byte) (int64, error) {
+	var h int64
+	buf := bytes.NewBuffer(data)
+	err := binary.Read(buf, binary.BigEndian, &h)
+	return h, errors.Trace(err)
+}
+
+type memTableReader struct {
+	baseExecutor
+	table         *model.TableInfo
+	columns       []*model.ColumnInfo
+	kvRanges      []kv.KeyRange
+	desc          bool
+	conditions    []expression.Expression
+	addedRows     [][]types.Datum
+	retFieldTypes []*types.FieldType
+	colIDs        map[int64]int
+	// cache for decode handle.
+	handleBytes []byte
+}
+
+func buildMemTableReader(us *UnionScanExec, tblReader *TableReaderExecutor) *memTableReader {
+	kvRanges := tblReader.kvRanges
+	colIDs := make(map[int64]int)
+	for i, col := range tblReader.columns {
+		colIDs[col.ID] = i
+	}
+
+	return &memTableReader{
+		baseExecutor:  us.baseExecutor,
+		table:         tblReader.table.Meta(),
+		columns:       us.columns,
+		kvRanges:      kvRanges,
+		desc:          us.desc,
+		conditions:    us.conditions,
+		addedRows:     make([][]types.Datum, 0, len(us.dirty.addedRows)),
+		retFieldTypes: us.retTypes(),
+		colIDs:        colIDs,
+		handleBytes:   make([]byte, 0, 16),
+	}
+}
+
+func (m *memTableReader) getMemRows() ([][]types.Datum, error) {
```

Comment:
How about making `memXXXReader` lazy?
We don't need to decode all rows at the beginning of the reading in some cases like `select col from t limit 1`.
Maybe we can amend `getMemRows` to `getOneRow` and decode one row when we need it.

Reply:
Great, Maybe we can do this in the next PR. Already add a TODO here.